### PR TITLE
Adding VEP splitting logic into split_multi_hts

### PR DESCRIPTION
**Pull request by konradjk**

### CODE REVIEW

File: python/hail/methods/statgen.py [hunk 1]
```diff
@@ -1604,10 +1606,23 @@ def split_multi_hts(ds, keep_star=False, left_aligned=False) -> MatrixTable:
 
     """
 
+    row_fields = set(ds.row)
     entry_fields = set(ds.entry)
     split = split_multi(ds, keep_star=keep_star, left_aligned=left_aligned)
+    update_rows_expression = {}
     update_entries_expression = {}
 
+    if vep_root in row_fields:
```
Comment (tpoterba) [inline python/hail/methods/statgen.py]: will this work for all VEPs or just human VEP? with the recent changes, we should try to be flexible.

Also I think this is broken -- a_index is a field, not a method

File: python/hail/methods/statgen.py [hunk 2]
```diff
@@ -1635,7 +1650,9 @@ def split_multi_hts(ds, keep_star=False, left_aligned=False) -> MatrixTable:
     if 'PID' in entry_fields:
         update_entries_expression['PID'] = split.PID
 
-    return split.annotate_entries(**update_entries_expression).drop('old_locus', 'old_alleles')
+    return split.annotate_rows(
+        **update_rows_expression).annotate_entries(
```
Comment (tpoterba) [inline python/hail/methods/statgen.py]: also this will invalidate the annotate entries

Comment (tpoterba) [reply python/hail/methods/statgen.py]: need to use `_annotate_all` I think

Comment (konradjk) [reply python/hail/methods/statgen.py]: nifty!

File: python/hail/methods/statgen.py (hunk 1 above)
Comment (konradjk) [reply python/hail/methods/statgen.py]: yep, good catch. some old code copied from back when it was

Comment (konradjk) [reply python/hail/methods/statgen.py]: and i _believe_ all VEPs add these 4 fields, but i could be run. are you guys testing mouse VEP? or has anyone over there run it and can show me what the results look like?

**Comment by danking** [comment]
I'm reassigning to tim because I don't think I understand VEP enough to know when this looks correct.

File: python/hail/methods/statgen.py [hunk 3]
```diff
@@ -1604,10 +1607,23 @@ def split_multi_hts(ds, keep_star=False, left_aligned=False) -> MatrixTable:
...
+    row_fields = set(ds.row)
     entry_fields = set(ds.entry)
     split = split_multi(ds, keep_star=keep_star, left_aligned=left_aligned)
+    update_rows_expression = {}
     update_entries_expression = {}
 
+    if vep_root in row_fields:
+        update_rows_expression[vep_root] = split[vep_root].annotate(
+            intergenic_consequences=split[vep_root].intergenic_consequences.filter(
+                lambda csq: csq.allele_num == split.a_index),
```
Comment (tpoterba) [inline python/hail/methods/statgen.py]: can make this a little nicer:
```python
        update_rows_expression[vep_root] = split[vep_root].annotate(**{
            x: split[vep_root][x].filter(lambda csq: csq.allele_num == split.a_index) 
            for x in ('intergenic_consequences', ...)})
```

File: python/hail/methods/statgen.py [hunk 4]
```diff
@@ -1596,6 +1597,8 @@ def split_multi_hts(ds, keep_star=False, left_aligned=False) -> MatrixTable:
         If ``True``, variants are assumed to be left
         aligned and have unique loci. This avoids a shuffle. If the assumption
         is violated, an error is generated.
+    vep_root : :obj:`str`
```
Comment (tpoterba) [inline python/hail/methods/statgen.py]: can you put a sentence in the docs about what this does?

Comment (konradjk) [reply python/hail/methods/statgen.py]: Is this the way you want this? I was toying with adding an expression, but that would be different from all the others (GQ, etc). I wonder if it's even worth an argument

Comment (tpoterba) [reply python/hail/methods/statgen.py]: passing in an expression is no easier than doing an `annotate_rows`, which would be equivalent. I think this is nice and simple

**Comment by konradjk** [comment]
I will do a quick sanity check on this before dismissing the review and requesting anew. Right now no tests cover this, but @jbloom22's framework could if we really wanted (annotate a multi-allelic, split it using this, and assert that each allele has at least one annotation)

**Comment by konradjk** [comment]
Ok a spot check looks sensible using:
```
mt = hl.utils.range_matrix_table(1, 1).key_rows_by(locus=hl.locus('20', 31022441), alleles='AGG,AGGG,AG,TGG,A'.split(','))
mt = hl.vep(mt, vep_config)
mt.transmute_rows(vep=mt.vep.transcript_consequences).rows().explode('vep').show(20)
```
returns:
```
+---------+---------------+-------------------------------+----------------+
| row_idx | locus         | alleles                       | vep.allele_num |
+---------+---------------+-------------------------------+----------------+
|   int32 | locus<GRCh37> | array<str>                    |          int32 |
+---------+---------------+-------------------------------+----------------+
|       0 | 20:31022441   | ["AGG","AGGG","AG","TGG","A"] |              3 |
|       0 | 20:31022441   | ["AGG","AGGG","AG","TGG","A"] |              3 |
|       0 | 20:31022441   | ["AGG","AGGG","AG","TGG","A"] |              3 |
|       0 | 20:31022441   | ["AGG","AGGG","AG","TGG","A"] |              3 |
|       0 | 20:31022441   | ["AGG","AGGG","AG","TGG","A"] |              3 |
|       0 | 20:31022441   | ["AGG","AGGG","AG","TGG","A"] |              1 |
|       0 | 20:31022441   | ["AGG","AGGG","AG","TGG","A"] |              1 |
|       0 | 20:31022441   | ["AGG","AGGG","AG","TGG","A"] |              1 |
|       0 | 20:31022441   | ["AGG","AGGG","AG","TGG","A"] |              1 |
|       0 | 20:31022441   | ["AGG","AGGG","AG","TGG","A"] |              1 |
|       0 | 20:31022441   | ["AGG","AGGG","AG","TGG","A"] |              2 |
|       0 | 20:31022441   | ["AGG","AGGG","AG","TGG","A"] |              2 |
|       0 | 20:31022441   | ["AGG","AGGG","AG","TGG","A"] |              2 |
|       0 | 20:31022441   | ["AGG","AGGG","AG","TGG","A"] |              2 |
|       0 | 20:31022441   | ["AGG","AGGG","AG","TGG","A"] |              2 |
|       0 | 20:31022441   | ["AGG","AGGG","AG","TGG","A"] |              4 |
|       0 | 20:31022441   | ["AGG","AGGG","AG","TGG","A"] |              4 |
|       0 | 20:31022441   | ["AGG","AGGG","AG","TGG","A"] |              4 |
|       0 | 20:31022441   | ["AGG","AGGG","AG","TGG","A"] |              4 |
|       0 | 20:31022441   | ["AGG","AGGG","AG","TGG","A"] |              4 |
+---------+---------------+-------------------------------+----------------+
```
Then with this jar (`--jar gs://hail-ci-0-1/87727687463212ce41eb1b40631ee9faf65c4324/39701f053c8affb1b7534f20304fb11fe7c22b48/artifacts/hail-all-spark.jar --zip gs://hail-ci-0-1/87727687463212ce41eb1b40631ee9faf65c4324/39701f053c8affb1b7534f20304fb11fe7c22b48/artifacts/hail-python.zip`) split works properly:
```
mt = hl.split_multi_hts(mt)
mt.transmute_rows(vep=mt.vep.transcript_consequences).rows().explode('vep').show(20)
```
returns:
```
+---------+---------------+-------------+----------------+-----------------+
| row_idx | locus         | alleles     | vep.allele_num | vep.amino_acids |
+---------+---------------+-------------+----------------+-----------------+
|   int32 | locus<GRCh37> | array<str>  |          int32 | str             |
+---------+---------------+-------------+----------------+-----------------+
|       0 | 20:31022441   | ["A","AG"]  |              1 | G/GX            |
|       0 | 20:31022441   | ["A","AG"]  |              1 | G/GX            |
|       0 | 20:31022441   | ["A","AG"]  |              1 | NA              |
|       0 | 20:31022441   | ["A","AG"]  |              1 | NA              |
|       0 | 20:31022441   | ["A","AG"]  |              1 | NA              |
|       0 | 20:31022441   | ["A","T"]   |              3 | G               |
|       0 | 20:31022441   | ["A","T"]   |              3 | G               |
|       0 | 20:31022441   | ["A","T"]   |              3 | NA              |
|       0 | 20:31022441   | ["A","T"]   |              3 | NA              |
|       0 | 20:31022441   | ["A","T"]   |              3 | NA              |
|       0 | 20:31022441   | ["AG","A"]  |              2 | G/X             |
|       0 | 20:31022441   | ["AG","A"]  |              2 | G/X             |
|       0 | 20:31022441   | ["AG","A"]  |              2 | NA              |
|       0 | 20:31022441   | ["AG","A"]  |              2 | NA              |
|       0 | 20:31022441   | ["AG","A"]  |              2 | NA              |
|       0 | 20:31022441   | ["AGG","A"] |              4 | G/X             |
|       0 | 20:31022441   | ["AGG","A"] |              4 | G/X             |
|       0 | 20:31022441   | ["AGG","A"] |              4 | NA              |
|       0 | 20:31022441   | ["AGG","A"] |              4 | NA              |
|       0 | 20:31022441   | ["AGG","A"] |              4 | NA              |
+---------+---------------+-------------+----------------+-----------------+
```

**Pull request merged**